sw/device/tests/alert_handler_lpg_sleep_mode_pings.c

```diff
@@ -0,0 +1,425 @@
+  // We don't expect test_step_cnt to 256 (flash mem is filled with all zeros)
+  // for this test. If it is 256, we just initialize the counters/NVM-fields
+  // again.
+  if (test_step_cnt == 256) {
+    flash_ctrl_testutils_counter_init_zero(&flash_ctrl, kCounterTestSteps);
+    flash_ctrl_testutils_counter_init_zero(&flash_ctrl, kCounterTestPhase);
+    test_step_cnt = flash_ctrl_testutils_counter_get(kCounterTestSteps);
+    rnd_num_iterations = flash_ctrl_testutils_counter_get(kCounterTestPhase);
+  }
+
+  // If this is the first iteration,
+  // set num_iterations to a random value between 8 and 32
+  if (rnd_num_iterations == 0) {
+    // 8 <= num_iters <= 32
+    rnd_num_iterations = rand_testutils_gen32_range(8, 32);
+    flash_ctrl_testutils_counter_set_at_least(&flash_ctrl, kCounterTestPhase,
+                                              rnd_num_iterations);
+    // LOG_INFO("phase1-num-iters = %d", rnd_num_iterations);
+  }
+
+  // Test phase #1: ping-timeout = 256
+  // No ping_timeout_fail alert shall be fired
+  // To ensure that the ping mechanism won't send spurious failure.
+  while (test_step_cnt < rnd_num_iterations) {
+    // Configure and lock the alerts
+    CHECK_DIF_OK(
+        dif_alert_handler_is_ping_timer_locked(&alert_handler, &is_locked));
+    if (!is_locked) {
+      alert_handler_config(/*ping_timeout=*/256);
+    }
+
+    // Power-on reset
+    if (pwrmgr_testutils_is_wakeup_reason(&pwrmgr, 0)) {
+      LOG_INFO("POR reset");
+      // Set  the AON timer to send a wakeup signal in ~150us.
+      // Go to deep sleep mode
+      aon_timer_testutils_wakeup_config(&aon_timer, 30);
+      enter_low_power(/*deep_sleep=*/(test_step_cnt % 2));
+    } else { /*wakeup reset*/
+      // Check the wakeup reason and do the cleanup
+      check_wakeup_reason();
+      cleanup_wakeup_src();
+
+      // After the wakeup, wait enough, and
+      // check that no ping-timeout-fail alert is fired
+      wait_enough_for_alert_ping();
+      CHECK_DIF_OK(dif_alert_handler_local_alert_is_cause(
+          &alert_handler, kDifAlertHandlerLocalAlertAlertPingFail, &is_cause));
+      CHECK(!is_cause,
+            "Expected response is ping_timeout_fail= 0! But we got "
+            "ping_timeout_fail = 1");
+
+      // Increment the test_step counter
+      flash_ctrl_testutils_counter_increment(&flash_ctrl, kCounterTestSteps);
+      test_step_cnt = flash_ctrl_testutils_counter_get(kCounterTestSteps);
+      // LOG_INFO("Test %d beginn", test_step_cnt);
+      // Set  the AON timer to send a wakeup signal in ~150us.
+      aon_timer_testutils_wakeup_config(&aon_timer, 30);
+      // Enter the normal sleep or deep sleep mode
+      enter_low_power(/*deep_sleep=*/(test_step_cnt % 2));
+      // enter_low_power(/*deep_sleep=*/rand_testutils_gen32_range(0,1));
+    }
+  }
+
+  // Test phase #2: ping_timeout = 2
+  // ping_timeout_fail alert shall be fired.
+  // To ensure that ping mechanism will continue to send out pings after wakeup
```

Comment:
@cindychip, you are right: 
- we need to reconfigure the alert_handler while waking up from the deep sleep mode ([CONTROL.MAIN_PD_N = 0](https://docs.opentitan.org/hw/ip/pwrmgr/doc/#Reg_control)) 
- As the normal sleep just disables the clocks, we don't need to reconfigure the alerts while waking up from the normal sleep mode.

There is some room for interpretation in the test spec. As far as I understand, the main purpose of this test is "to verify that the alert_ping mechanism works as expected after a bunch of deep/normal sleep/wakeup events". Please correct me if I am wrong:
- Because of [this line](https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/data/chip_testplan.hjson#L1465), after waking up from the normal sleep, I first clear the previous ping-timeout-fail event and then wait for a new ping-timeout-fail event.
- Again, because of the same line of the test description, after waking up from the deep sleep, I reconfigure the alert handler first and then apply the same steps as the normal wakeup case.

I am okay with your suggestion too. Maybe we can set up an offline discussion for clarification of the test description.

Reply:
I think @bilgiday's interpretation is fine - we want to make sure that the ping mechanism keeps on working after coming back from sleep scenarios. The interesting case is light sleep, since that one is supposed to retain the alert handler configuration. The deep sleep one is also good to have but it is slightly less interesting since the alert handler is reset and reconfigured.

I have however a few suggestions on how to tweak the sequence below.

Reply:
Ok, thanks Michael, I guess what I am concerned about should be covered in chip_sw_alert_handler_lpg_sleep_mode_alerts :)
Thank you both for the explanation